src/libraries/Microsoft.PowerFx.Interpreter/Functions/Mutation/PatchFunction.cs

```diff
@@ -0,0 +1,267 @@
+            Contracts.AssertValue(args);
+            Contracts.AssertAllValues(args);
+            Contracts.AssertValue(argTypes);
+            Contracts.Assert(args.Length == argTypes.Length);
+            Contracts.AssertValue(errors);
+
+            var isValid = CheckInvocation(args, argTypes, errors, out returnType, out nodeToCoercedTypeMap);
+
+            DType dataSourceType = argTypes[0];
+            DType retType = DType.EmptyRecord;
+
+            for (var i = 1; i < args.Length; i++)
+            {
+                DType curType = argTypes[i];
+
+                if (!curType.IsRecord)
+                {
+                    errors.EnsureError(args[i], TexlStrings.ErrNeedRecord);
+                    isValid = false;
+                    continue;
+                }
+
+                var isSafeToUnion = true;
+
+                foreach (var typedName in curType.GetNames(DPath.Root))
+                {
+                    DName name = typedName.Name;
+
+                    if (!dataSourceType.TryGetType(name, out DType dsNameType))
+                    {
+                        dataSourceType.ReportNonExistingName(FieldNameKind.Display, errors, name, args[i]);
+                        isValid = isSafeToUnion = false;
+                        continue;
+                    }
+                }
+
+                if (isValid && SupportsParamCoercion && !dataSourceType.Accepts(curType))
+                {
+                    if (!curType.TryGetCoercionSubType(dataSourceType, out DType coercionType, out var coercionNeeded))
+                    {
+                        isValid = false;
+                    }
+                    else
+                    {
+                        if (coercionNeeded)
+                        {
+                            CollectionUtils.Add(ref nodeToCoercedTypeMap, args[i], coercionType);
+                        }
+
+                        retType = DType.Union(retType, coercionType);
+                    }
+                }
+                else if (isSafeToUnion)
+                {
+                    retType = DType.Union(retType, curType);
+                }
+            }
+
+            return isValid;
+        }
+
+        public async Task<FormulaValue> InvokeAsync(FormulaValue[] args, CancellationToken cancel)
+        {
+            // If any args are error, propagate up.
+            foreach (var arg in args)
+            {
+                if (arg is ErrorValue)
+                {
+                    return arg;
+                }
+            }
+
+            if (args[1] is BlankValue)
+            {
+                return args[1];
+            }
+
+            var argFields = CreateRecordFromArgsDict(args, 1, out ErrorValue errorValue);
+
+            if (errorValue != null)
+            {
+                return errorValue;
+            }
+
+            var arg1 = args[1] as RecordValue;
+
+            return arg1.UpdateFields(argFields);
```

Comment:
the more defensive coding would be:

 var arg1 = (RecordValue) args[1]; 

Before, if for some reason it's not a RecordValue, then 'as' operator returns null, and we'd get a null ref on line 264 -and null refs an be hard to track down. 

Whereas the cast will generated an explicit cast exception.